src/Course/List.hs

```diff
@@ -302,23 +390,66 @@ question_List_8 = error "TODO"
 --
 -- Implement this function using pattern matching.
 filter :: (a -> Bool) -> List a -> List a
-filter = error "todo: Course.List#filter"
+filter _ Nil = Nil
+filter f (x :. xs) =
+  let
+    rest = filter f xs
+  in
+    if (f x) then (x :. rest) else rest 
 
 -- | Question List 9
 --
 --   (a) Implement `map` using `foldRight`.
 --   (b) Implement `filter` using `foldRight`.
+--
+-- > map' (3+) (1 :. 2 :. 3 :. Nil)
+-- [4,5,6]
+--
+-- > filter' odd (1 :. 2 :. 3 :. Nil)
+-- [1,3]
+map' :: (a -> b) -> List a -> List b
+map' f xs = foldRight step init list
+  where
+    step = \next acc -> (f next) :. acc
+    init = Nil
+    list = xs
+
+filter' :: (a -> Bool) -> List a -> List a
+filter' f xs = foldRight step init list
+  where
+    step = \next acc -> if (f next) then (next :. acc) else acc
+    init = Nil
+    list = xs
+
 question_List_9 :: (String, String)
-question_List_9 = error "TODO"
+question_List_9 =
+  ( "see `map'`"
+  , "see `filter'`"
+  )
 
 -- | Append two lists to a new list.
 --
 -- > (1 :. 2 :. 3 :. Nil) ++ (4 :. 5 :. 6 :. Nil)
 -- [1,2,3,4,5,6]
 --
 -- Implement this function using `foldRight`.
+--
+-- Daniel shared a great technique on how to set myself up for
+-- thinking clearly about folds. Start by creating scaffolding using a `where` block
+-- to stub out the pieces of the fold. This allows for better isolation of concepts
+-- to be filled out, rather than trying to understand everything at once.
+-- (++) xs ys = foldRight step init list
+--   where
+--     step next acc = undefined
+--     init = undefined
+--     list = undefined
 (++) :: List a -> List a -> List a
-(++) = error "todo: Course.List#(++)"
+(++) xs ys = foldRight (:.) ys xs
```

Comment:
excellent! when you use `ys` as in `init`, `(++)` only needs to pattern match its first argument. it never needs to pattern match `ys` :-D